### PR TITLE
Add regression test for #29723

### LINKED ISSUE #29723: Variables moved from in match guards are still accessible in other match arms.
The following code segfaults at runtime on stable, beta, and nightly:

``` rust
fn main() {
    let x = 10;
    let s = Custom { val: Vec::new() };

    match x {
        10 if test(s) => println!("oops"),
        _ => println!("eek {:?}", s.val),
    }
}

#[derive(Debug)]
pub struct Custom {
    val: Vec<u8>,
}

impl Drop for Custom {
    fn drop(&mut self) { println!("Dropping"); }    
}

#[allow(unused_variables)]
fn test(s: Custom) -> bool { false }
```

### PR DESCRIPTION
**Pull request by Manishearth**
cc #29723

r? @nikomatsakis

### CODE REVIEW

File: src/test/ui/issue-29723.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    let s = String::new();
+    let _s = match 0 {
+        0 if { drop(s); false } => String::from("oops"),
+        _ => {
+            // This should trigger an error,
+            // s could have been moved from.
+            s
+        }
+    };
+}
```
Comment (kennytm) [inline src/test/ui/issue-29723.rs]:
```
[00:04:49] tidy error: /checkout/src/test/ui/issue-29723.rs: missing trailing newline
```

**Comment by Manishearth** [comment]
fixed

**Comment by nikomatsakis** [comment]
[missing main function](https://travis-ci.org/rust-lang/rust/builds/330689203#L1587)

**Comment by nikomatsakis** [comment]
r=me once travis is green -- probably just change `foo` to `main`, no?

**Comment by nikomatsakis** [comment]
I did it myself :)

**Pull request merged**